### DISCUSSION TITLE
Unable to Connect to traced_consumer Socket in Perfetto on Android

### DISCUSSION THREAD

**Discussion by EndSmile** [google/perfetto#964] [Q&A] [answered]
I created a demo application to integrate Perfetto tracing. Here’s what I did:

1. Placed perfetto.h and perfetto.cc in my cpp
2. Created a JNI interface to call the main method from example_startup_trace.cc
When I ran the Perfetto command via ADB and invoked example_startup_trace, the following logs were printed in the command line:

```
I/perfetto( 1495):    ftrace_procfs.cc:335 enabled ftrace in /sys/kernel/tracing/
D/perfetto(24934):       perfetto.cc:38836 Reconnecting backend 1 for startup tracing
D/perfetto(24934):       perfetto.cc:62995 Tracing service connection failure
D/perfetto(24934):       perfetto.cc:38858 Setting up data source com.example.startup_trace for startup tracing with target buffer reservation 1
D/perfetto(24934):       perfetto.cc:36615 Producer connected
D/perfetto(24934):       perfetto.cc:62072 Tracing service connection failure
E/perfetto(24934):       perfetto.cc:36863 Unable to connect to the system tracing service as a consumer. On Android, use the "perfetto" command line tool instead to start system-wide tracing sessions
D/perfetto(24934):       perfetto.cc:36615 Producer connected
I/perfetto(24934): le_startup_trace.cc:107 Trace written in /data/user/0/com.shopee.videoperf/cache/example_startup_trace.pftrace file. To read this trace in text form, run `./tools/traceconv text /data/user/0/com.shopee.videoperf/cache/example_startup_trace.pftrace`
```
Additionally, I observed the following log in my Android logcat:

`type=1400 audit(0.0:10007): avc: denied { write } for name="traced_consumer" dev="tmpfs" ino=21663 scontext=u:r:untrusted_app:s0:c117,c257,c512,c768 tcontext=u:object_r:traced_consumer_socket:s0 tclass=sock_file permissive=0 app=com.xxx.videoperf`

**Questions**
1. If the connection to traced_consumer fails, why are these logs still printed by the Perfetto command? 
2. The traced_consumer socket appears to be inaccessible to normal applications due to permissions (avc: denied). Does this mean Perfetto is not designed to be used directly by regular applications, and is its usage limited to system services or privileged contexts?

**Comment by LalitMaganti** [comment, marked as answer]
There are multiple parts to your question so let me go through one by one:
> 1) When I ran the Perfetto command via ADB and invoked example_startup_trace, the following logs were printed in the command line

If you are trying to use the consumer side of Perfetto from your app, why are you using the `adb shell perfetto` command? The two things are mutually exclusive: either you are using perfetto_cmd as your consumer or your own app as your consumer. You cannot be doing both.

> If the connection to traced_consumer fails, why are these logs still printed by the Perfetto command?

The answer to this needs me to understand what you are doing above.

> The traced_consumer socket appears to be inaccessible to normal applications due to permissions (avc: denied). Does this mean Perfetto is not designed to be used directly by regular applications, and is its usage limited to system services or privileged contexts?

The consumer side of Perfetto is *not* designed for direct use by apps. The producer side is accessible to everyone. I'd suggest checking out https://perfetto.dev/docs/design-docs/security-model which goes into the purpose of each side and why the security model is the way it is.

Note the following though:
1) In local/lab settings, you can use `adb shell perfetto` to collect traces without needing your app to talk to the consumer side: that's totally supported and normal use of Perfetto.
2) In production usecases (i.e. in the field), in Android V+, we introduced a new platform API called ProfilingManager (https://developer.android.com/reference/android/os/ProfilingManager) which provides a controlled version of the Perfetto consumer API for apps to request traces. It won't give you access to have arbitrary configs (for security and system health reasons) *but* it will probabilistically allow you to request traces from the system at certain times.

**Reply by EndSmile** [reply]
Thank you for your answer. I was indeed unfamiliar with the producer-consumer model before. My current question is that I don't know how to use the producer in Perfetto independently. The reason is that the ATrace API on Android lacks sufficient functionality, so I want to use Perfetto's API directly. However, it seems that the examples don't cover this scenario.

Additionally, if I run a consumer within the app, can it only receive events produced by itself, or can it also capture ATrace events?

**Reply by LalitMaganti** [reply]
> current question is that I don't know how to use the producer in Perfetto independently.

For advanced users who are willing to gets hands dirty you can look at https://perfetto.dev/docs/instrumentation/track-events. But for most users we still recommend using atrace. Yes it's limited but the Android has survived with using it for 15+ years, it can survive a couple more. 

We are working on building proper platform APIs for wrapping this to make things easier but that will take some time. 

> Additionally, if I run a consumer within the app, can it only receive events produced by itself, or can it also capture ATrace events?

Only those generated by the app itself with the Perfetto APIs only. You will not get any system data.

**Reply by EndSmile** [reply]
Sorry, I still cannot see the trace in Perfetto UI. I read the documentation and wrote the following example code. The adb command prints:

```
D/perfetto(12268):       perfetto.cc:36615 Producer connected  
D/perfetto(12268):       perfetto.cc:36615 Producer connected  
```
It seems that the producer has successfully connected, but my adb perfetto command still doesn't receive any trace. I tried adjusting some configurations, such as modifying the backends, but it had no effect. I'm not sure where the problem lies. Is this usage not supported?

```
#include <chrono>
#include <fstream>
#include <thread>
#include <jni.h>

#include <perfetto.h>
// The set of track event categories that the example is using.
PERFETTO_DEFINE_CATEGORIES(
        perfetto::Category("rendering")
                .SetDescription("Rendering and graphics events"),
        perfetto::Category("network.debug")
                .SetTags("debug")
                .SetDescription("Verbose network events"),
        perfetto::Category("audio.latency")
                .SetTags("verbose")
                .SetDescription("Detailed audio latency metrics"));

namespace {

void InitializePerfetto() {
  perfetto::TracingInitArgs args;
  args.backends = perfetto::kSystemBackend;

  perfetto::Tracing::Initialize(args);
  perfetto::TrackEvent::Register();
}

void DrawPlayer(int player_number) {
  TRACE_EVENT("rendering", "DrawPlayer", "player_number", player_number);
  // Sleep to simulate a long computation.
  std::this_thread::sleep_for(std::chrono::milliseconds(500));
}

}  // namespace

int main() {
  InitializePerfetto();
  DrawPlayer(1);
  return 0;
}
```
----- 
Let me sync some background knowledge:
The reason I didn’t directly use the `ATrace` API is that it doesn’t directly support instant events or passing `args` in traces. I want to build a trace framework for internal app use, which can use perfetto to capture traces offline and support converting traces to JSON format for uploading to ES. The JSON format should support `args` in traces.

For example:
```
[
    {
        "cat": "example.category",
        "name": "ExampleTraceStart",
        "ph": "I",
        "tid": 1,
        "ts": 14,
        "args": {
            "detail": "test"
        }
    }
]
```

**Reply by LalitMaganti** [reply]
> Is this usage not supported?

What does your config look like? Is track_event enabled there?

**Reply by EndSmile** [reply]
Thank you very much, the problem has been solved.
